Add real-time comments to Reader full-post view

mhsdef opened the pull request:
We're getting our Calypso ["sea legs"](https://www.merriam-webster.com/dictionary/sea%20legs) back under us. And very open to detailed integration input as this gets off the ground. 

1. Make it work. 
2. Make it beautiful. <--here.
3. Make it fast (if you need to, "90% of the time" step two has you covered).

#### Changes proposed in this Pull Request

* Add [Phoenix](https://github.com/phoenixframework/phoenix)-based websocket middleware.
* Add real-time comments to Reader full-post view behind a feature flag.
* Disable Reader comment polling while we explore the space.

#### Testing instructions

1. Apply D40881-code to your sandbox.
2. Fire up this branch locally or [hit calypso.live](https://calypso.live/?branch=try/realtime-comments).
3. Go to a full post view in the Reader _on a public post_. View in one window and comment in another. Comment from traditional WP.com. Comment from Notifs.

You should see the Reader update real-time in all cases.

mhsdef committed:
Add Lasagna websocket middleware (first pass)

blowery commented:
Amazing!

What's lasagna?

hewsut commented:
A backend piece `try`, over here: https://github.com/Automattic/lasagna

blowery commented:
@hewsut cooooool

mhsdef committed:
Step forward

mhsdef committed:
Sync to latest changes

mhsdef committed:
One other file

mhsdef committed:
Fix topics

mhsdef committed:
Check in a bunch of stuff

a8ci18n commented on client/blocks/presence-button/index.jsx:
```diff
@@ -0,0 +1,79 @@
+
+			{ showLabel && (
+				<span className="presence-button__label-status">
+					{ translate( 'Viewing', 'Viewing', {
```
:information_source: This string doesn't exist elsewhere without a **context** (or with a different one). You can probably drop the context. If needed, you can use a translator comment instead to clarify meaning.

mhsdef committed:
Remove Presence for now

mhsdef committed:
Fix logging path

mhsdef committed:
Remove Presence stragglers

mhsdef committed:
More Presence stragglers

mhsdef committed:
More cleanup

mhsdef committed:
Merge branch 'master' into try/realtime-comments

mhsdef committed:
Iterate

hewsut commented:
It's clear we will have at least a couple more followup PRs to polish UX around things like the stream views, comment count in the post action button, etc. But if I did this thing right what we're doing is behind a feature flag.

So I think we're ready for any and all feedback here with intent to merge to master.

blowery commented on client/state/lasagna/socket.js:
```diff
@@ -0,0 +1,50 @@
+/**
+ * External Dependencies
+ */
+import { Socket } from 'phoenix';
```
could we dynamically import this when we actually need it?

blowery replied:
I ask because this is bloating the main entry point, which nearly everything loads, by about 12KB (compressed). If we can load phoenix on demand, we should be able to get back a chunk of that.

hewsut replied:
Yes! How? 😅 

/me googling and grepping our JS code bases furiously.

hewsut replied:
Try to push the `import { socket } from 'state/lasagna/socket'` import back out of `client/sections-middleware.js`?

blowery replied:
dynamically import it here instead of statically. remove this line and add 
```
import('phoenix').then( { Socket } => { 
 ...
})
```
down below where you use the Socket

blowery replied:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#Dynamic_Imports

hewsut replied:
Thanks! ICFY still running as I comment but I think we've got that in place successfully now.

mhsdef committed:
Push connection management back into the middleware

Duh?

mhsdef committed:
Add Reader full post view state subtree

We need a way to query what is onscreen in a Reader full post view if a user hits a reader link directly. The use case is joining related channel(s) after the websocket connects.

mhsdef committed:
Use new full view post state logic

mhsdef committed:
Switch to dynamic load of Phoenix.js

mhsdef committed:
Additional cleanup

unDemian committed:
Merge branch 'master' into try/realtime-comments

mhsdef committed:
Merge branch 'master' into try/realtime-comments

mhsdef committed:
Remove stray addition

To come in follow up PR

mhsdef committed:
Add config.isEnabled check back for lasagna

unDemian committed:
Merge branch 'master' into try/realtime-comments

lynnbgriffith reviewed (changes requested):
I believe a `webpackChunkName` comment should ensure the dynamic `phoenix` import is always included in a separate, async-loaded chunk.  See https://webpack.js.org/guides/code-splitting/#dynamic-imports

mhsdef committed:
Merge branch 'master' into try/realtime-comments

mhsdef committed:
Add webpackchunkname to dynamic import

hewsut commented:
@griffbrad I thought I had this nailed in https://github.com/Automattic/wp-calypso/pull/40396/commits/7277061853bb594477ae267a66c551e64b9acd19 after your feedback but ICFY still seems to report +12 KB. 

FWIW I can see it coming down as a separate JS chunk now:

![Screen Shot 2020-04-07 at 12 54 32 PM](https://user-images.githubusercontent.com/2006764/78697540-f2c7d500-78ce-11ea-83c7-0bf3cd58d618.png)

mhsdef committed:
To be handled in a subsequent PR

griffbrad commented:
@sgomes have you encountered this before?  Bundle size on entry-main is inflated in ICFY but the dependency in question is being loaded async in dev tools.

unDemian committed:
Merge branch 'master' into try/realtime-comments

unDemian commented:
My guess is the `12kb` is not from loading Phoenix's socket which seems to be correctly lazy loaded and only be around `6kb` (https://d.pr/i/7WVJiT).
I think it might be the actual lasagna middleware and its dependencies which is always loaded

https://github.com/Automattic/wp-calypso/blob/0a3aa2238a7fc2d9129233314119a579e560516d/client/state/index.js#L71

Subsequent PRs built on top of this one https://github.com/Automattic/wp-calypso/pull/40720 https://github.com/Automattic/wp-calypso/pull/40854 show an increase since they add more code tied to that middleware.

- Here is an identical branch which doesn't load the middleware in state https://github.com/Automattic/wp-calypso/pull/40883, **entry-main is almost untouched**.

- Here is an identical branch which loads the middleware code but doesn't load phoenix at all
https://github.com/Automattic/wp-calypso/pull/40889, **entry-main is affected by the same amount**.

I'm not quite sure how we can fix this, I guess we would need some kind of async middleware registration if we want to conditionally load the middleware code only when needed, I bet there is some kind of way of doing that in Calypso 🤷‍♂️

sgomes commented:
Excellent research, @unDemian! I agree with your assessment; the middleware appears to be at fault here. I don't think there is currently any example of an asynchronously-loaded middleware anywhere in Calypso, but it should be possible, at least in theory. In practice, there may need to be some changes to the state layer in order to allow for it 😕 

Unfortunately, any middleware approach is prone to these sorts of issues, since it separates functionality from the place where it's used, and makes it available everywhere. I've been looking at similar issues with analytics (which has its own middleware), as I'm trying to avoid the whole thing being loaded every time, regardless of what's in use.

sirreal commented on client/state/lasagna/actions.js:
```diff
@@ -0,0 +1,2 @@
+export const socketConnected = () => ( { type: 'LASAGNA_SOCKET_CONNECTED' } );
+export const socketDisconnected = () => ( { type: 'LASAGNA_SOCKET_DISCONNECTED' } );
```
The action types are added to `client/state/action-types` but plain strings are used here.

unDemian replied:
That should be fixed in the next PR https://github.com/Automattic/wp-calypso/pull/40720

unDemian commented:
> Unfortunately, any middleware approach is prone to these sorts of issues, since it separates functionality from the place where it's used, and makes it available everywhere. I've been looking at similar issues with analytics (which has its own middleware), as I'm trying to avoid the whole thing being loaded every time, regardless of what's in use.

I guess having a "placeholder" middleware that allows dynamic add/remove would do the trick, something like https://github.com/pofigizm/redux-dynamic-middlewares maybe?

lezama commented:
@griffbrad @sgomes

Two questions (or three or four)!

- We are talking about just +3.7% increase, right? could we go ahead merge it and fix the middleware async loading later 😅?
- Other than bundle size, what do you think about the code and approach? Ready to merge? 😃

lynnbgriffith reviewed (approved):
OK with merging now and circling back on the perf concerns, yes.  Sergio and I are discussing more and he's graciously offered to help with the middleware perf.  The overall approach seems solid.

mhsdef committed:
Final adjustment

mhsdef committed:
Merge branch 'master' into try/realtime-comments

hewsut commented:
Thanks all. I'm going to merge then whenever all these tests clear. Andrei has three or four right behind this that will improve immediately.
